### ISSUE TITLE
Cannot open another page when a loop is running

### ISSUE THREAD

**Issue by lvvvmd** [miguelgrinberg/Flask-SocketIO#934]
Hi, the server works well when only 1 client is connected and doing some cpu-costing jod. But if I want to open another page in the browser at the same time, it takes much time to get the HTML page and cannot establish a socketio connection.  
Codes are here:

app.py
```
import gevent
from gevent import monkey; monkey.patch_all()

from flask import Flask, render_template, session, request,Response, json
from flask_socketio import SocketIO, emit
import time

from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor

def video_analysis(msg, s, sid):
      while True:
             do some cpu_costing job
             emit(...)

def set_stop():
      #do a easy job just like a[2] = False
      pass
app = Flask(__name__, template_folder='./')
app.config['SECRET_KEY'] = 'secret!'

socketio = SocketIO(app, async_handlers=True)
URL = ''
@app.route('/api/test', methods=['GET', 'POST'])
def index():
    print('index')
    return render_template('index.html')

@socketio.on('start_event')
def start_event(msg):
    set_stop(False, request.sid)
    video_analysis(msg['data'], socketio,request.sid)
  

@socketio.on('stop_event')
def stop_event(msg):
    print('stop_event')
    set_stop(True, request.sid)

@socketio.on('connect_event')
def connected_msg(msg):
    print(msg)
    print('connected_msg', msg['data'])
    print(request.sid)
    socketio.emit('server_response', {'data': request.sid}, room = request.sid)
    socketio.emit('server_response', {'data': 'I\'ve received.'}, room=request.sid)
    #socketio.start_background_task(video_analise, URL)
    #video_analysis(msg['data'], socketio)

@socketio.on('disconnect')
def test_disconnect():
    print('Client disconnected', request.sid)
    set_stop(True, request.sid)
```

uwsgi.ini
```
[uwsgi]
wsgi-file =/app/FER/app.py
callable = app
workers=1
http=:3031
gevent=500
http-websockets=true
logto = /app/FER/uwsgi_servers/1/log.txt
socket = /app/FER/uwsgi_servers/1/emoPrediction.sock
chmod-socket = 777
vacuum = true
die-on-term = true
lazy-app = true
async_handlers=true
buffer-size=32768
```

nginx.conf
```
events {
    worker_connections  1024;
}

http {
    include       mime.types;
    default_type  application/octet-stream;

    sendfile        on;
        error_log  /var/log/nginx/nginx_error.log  warn;
    keepalive_timeout  65;
    gzip  off;
    include uwsgi_params;
    uwsgi_param UWSGI_FILE    $app;
    uwsgi_param UWSGI_PYHOME    $virtualenv;

    server {
    listen 8080;
    server_name localhost;

    location / {
        include uwsgi_params;
        uwsgi_pass unix:///app/FER/uwsgi_servers/1/emoPrediction.sock;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        }
    }
}


```
index.html
```
<!DOCTYPE HTML>
<html>
<head>
    <title>Flask-SocketIO Test</title>
    <script type="text/javascript" src="//cdn.bootcss.com/jquery/3.1.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.bootcss.com/socket.io/2.1.1/socket.io.dev.js"></script>
    <script type="text/javascript" charset="utf-8">
    $(document).ready(function() {
        var socket = io.connect(window.location.host);
        //var socket = io.connect();
 
        socket.on('connect', function() {
            socket.emit('connect_event', {data: '11111111111111'});
        });

        socket.on('server_response', function(msg) {
                console.log(msg.data)
                        var newDate = new Date();
                        console.log(newDate.toGMTString());
                        $('#log').append('<br>' + $('<div/>').text('Received #' + ': ' + msg.data + newDate.toGMTString()).html());
        });

        socket.on('error_event', function(msg) {
                console.log(msg.data)
                var newDate = new Date();
                console.log(newDate.toGMTString());
                $('#log').append('<br>' + $('<div/>').text('Received #' + ': ' + msg.data + newDate.toGMTString()).html());
        
        });
        socket.on('emotion_event', function(msg) {
                console.log(msg.data)
                var newDate = new Date();
                console.log(newDate.toGMTString());
                $('#log').append('<br>' + $('<div/>').text('Received #' + ': ' + msg.data + newDate.toGMTString()).html());
                $('#log').append('<br>' + $('<div/>').text('Received #' + ': ' + msg.timestamp ).html());                                                      
        });
        socket.on('end_event', function(msg) {
                console.log(msg.data)
                var newDate = new Date();
                console.log(newDate.toGMTString());
                $('#log').append('<br>' + $('<div/>').text('Received #' + ': ' + msg.data + newDate.toGMTString()).html());
                                                        
        });


        $('form#emit').submit(function(event) {
                socket.emit('start_event', {data: $('#emit_data').val()});
                return false;
            });
        $('form#stop').submit(function(event) {
                socket.emit('stop_event', {data: "stop"});
                return false;
                                                           
        });
    });
    
    </script>   
</head>
<body>
    <h2>WebSokect</h2>
    <form id="emit" method="POST" action='#'>
        <input type="text" name="emit_data" id="emit_data" placeholder="Message">
        <input type="submit" value="Echo">
    </form>
    <form id="stop" method="POST" action='#'>
        <input type="submit" value="stop">
    </form>
    <div id='log'></div>
</body>
</html>
                  
```

**Comment by miguelgrinberg** [comment]
You can't do CPU heavy work in a gevent application, it was no designed for that.

**Comment by hextrim** [comment]
What would be a good solution to this problem? Using more workers,
executing the tasks by celery and on request rather than page load... Would
that work?

**Comment by miguelgrinberg** [comment]
Celery should work, but it isn't the only way. The important thing is to keep the CPU heavy stuff in a separate process from gevent.

**Comment by hextrim** [comment]
Can I assume the other option would be to spawn gevent where needed instead
of monkey patching all, and use threading ... ?

I thought gevent could manage concurrency? I must have misunderstood its
fundamental concept?

**Comment by lvvvmd** [comment]
> Celery should work, but it isn't the only way. The important thing is to keep the CPU heavy stuff in a separate process from gevent.

Thank you very much!
So do you have some advice for me to do such CPU heavy stuff in a saprate process where the socketio would emit something?

**Comment by lvvvmd** [comment]
I am trying like this:
```
from multiprocessing import Process

def video_analysis(socketio):
      while True:
            #doing some cpu heavy jos
            socketio.emit(...,room=...)

p = Process(target=video_analysis, args=(socketio, ))
p.start()
```
But the emit() did not work.

**Comment by miguelgrinberg** [comment]
@lvvvmd Right, the emit does not work because you moved it to the other process. This is actually tricky, If you can avoid emitting from the auxiliary process then that would simplify the problem. If not I recommend that you read the section of the documentation on emitting from external processes.

**Comment by miguelgrinberg** [comment]
> I thought gevent could manage concurrency? I must have misunderstood its
> fundamental concept?

You need to read about the specific method of cooperative concurrency used by gevent, which happens to be pretty similar to the model used by asyncio and also JavaScript.

**Comment by lvvvmd** [comment]
@miguelgrinberg 

Thank you very much! I will read that document carefully!
And I don't know why the cilent always disconnects after several seconds. Is there something wrong with my uwsgi or nginx config?

**Comment by miguelgrinberg** [comment]
@lvvvmd I don't know what's the reason for the disconnects. You can enable more verbose logging on the server to see if anything interesting appears there. If the disconnect always happens at a specific point in the life of the connection, then I would suspect a connection timeout either in uwsgi or in nginx, so you should check there.

**Comment by lvvvmd** [comment]
@miguelgrinberg The disconnect time is equal to the ping_interval parameters of the SocketIO object. So if I set ping_interval =20, the disconnect will happen after 20 secs and if set ping_interval = 20000 the disconnect will not happen. So I think this is the keypoint but this solution is not elegant. 
And I disabled all the nginx and uwsgi and run the raw app by the origin flask server, the disconnection happens just like before. So thw problem is the ping between the client and server.
But  I am still not sure what is the problem.
